libpod/container_internal.go

```diff
@@ -2738,3 +2740,130 @@ func (c *Container) update(resources *spec.LinuxResources, restartPolicy *string
+		}
+		return nil
+	}
+
+	if !c.state.StartupHCPassed {
+		c.state.StartupHCPassed = !wasEnabledHealthCheck
+		c.state.StartupHCSuccessCount = 0
+		c.state.StartupHCFailureCount = 0
+		if err := c.save(); err != nil {
+			return err
+		}
+		if wasEnabledHealthCheck {
+			c.recreateHealthCheckTimer(context.Background(), true, true)
+		}
+		return nil
+	}
+	return nil
+}
+
+func (c *Container) updateHealthCheck(newHealthCheckConfig interface{}, currentHealthCheckConfig interface{}, setConfigFunc func(config interface{})) error {
+	oldHealthCheckConfig := currentHealthCheckConfig
+	if !reflect.ValueOf(oldHealthCheckConfig).IsNil() {
+		if err := JSONDeepCopy(currentHealthCheckConfig, oldHealthCheckConfig); err != nil {
+			return err
+		}
+	}
+
+	setConfigFunc(newHealthCheckConfig)
+
+	if err := c.runtime.state.SafeRewriteContainerConfig(c, "", "", c.config); err != nil {
+		// Assume DB write failed, revert to old resources block
+		setConfigFunc(oldHealthCheckConfig)
+		return err
+	}
+
+	isStartup := false
+	changedTimer := false
+	if !reflect.ValueOf(oldHealthCheckConfig).IsNil() && !reflect.ValueOf(newHealthCheckConfig).IsNil() {
+		switch oldConfig := oldHealthCheckConfig.(type) {
+		case *manifest.Schema2HealthConfig:
+			newConfig := newHealthCheckConfig.(*manifest.Schema2HealthConfig)
+			changedTimer = oldConfig.Interval != newConfig.Interval
+		case *define.StartupHealthCheck:
+			newConfig := newHealthCheckConfig.(*define.StartupHealthCheck)
+			changedTimer = oldConfig.Interval != newConfig.Interval
+			isStartup = true
+		}
+	}
+
+	noHealthCheck := c.config.HealthCheckConfig != nil && slices.Contains(c.config.HealthCheckConfig.Test, "NONE")
+
+	if err := c.resetHealthCheckTimers(noHealthCheck, changedTimer, !reflect.ValueOf(oldHealthCheckConfig).IsNil(), isStartup); err != nil {
+		return err
+	}
+
+	checkType := "HealthCheck"
+	if isStartup {
+		checkType = "Startup HealthCheck"
+	}
+	logrus.Debugf("%s configuration updated for container %s", checkType, c.ID())
+
+	c.newContainerEvent(events.Update)
+	return nil
+}
+
+func (c *Container) updateGlobalHealthCheckConfiguration(globalOptions define.GlobalHealthCheckOptions) error {
+	oldHealthCheckOnFailureAction := c.config.HealthCheckOnFailureAction
+	oldHealthLogDestination := c.config.HealthLogDestination
+	oldHealthMaxLogCount := c.config.HealthMaxLogCount
+	oldHealthMaxLogSize := c.config.HealthMaxLogSize
+
+	if globalOptions.HealthCheckOnFailureAction != nil {
+		c.config.HealthCheckOnFailureAction = *globalOptions.HealthCheckOnFailureAction
+	}
+
+	if globalOptions.HealthMaxLogCount != nil {
+		c.config.HealthMaxLogCount = *globalOptions.HealthMaxLogCount
+	}
+
+	if globalOptions.HealthMaxLogSize != nil {
+		c.config.HealthMaxLogSize = *globalOptions.HealthMaxLogSize
+	}
+
+	if globalOptions.HealthLogDestination != nil {
+		c.valid = false
+		err := WithHealthCheckLogDestination(*globalOptions.HealthLogDestination)(c)
+		if err != nil {
+			return err
+		}
+		c.valid = true
```

Comment:
That is extremely ugly, setting a container to not valid like this is not sane if the pointer is shared acorrs several threads another container may see a invalid container and bail out. I don't think we that is a huge risk as we not really share these structs at the same time but still.

And calling CtrCreateOption function somehwere else is just not nice to read. Please just split out the validation bits into another function and then call it from here and WithHealthCheckLogDestination()